kernel/src/table_changes/scan.rs

```diff
@@ -0,0 +1,204 @@
+}
+
+/// Builder to read the `TableChanges` of a table.
+pub struct TableChangesScanBuilder {
+    table_changes: Arc<TableChanges>,
+    schema: Option<SchemaRef>,
+    predicate: Option<ExpressionRef>,
+}
+
+impl TableChangesScanBuilder {
+    /// Create a new [`TableChangesScanBuilder`] instance.
+    pub fn new(table_changes: impl Into<Arc<TableChanges>>) -> Self {
+        Self {
+            table_changes: table_changes.into(),
+            schema: None,
+            predicate: None,
+        }
+    }
+
+    /// Provide [`Schema`] for columns to select from the [`TableChanges`].
+    ///
+    /// A table with columns `[a, b, c]` could have a scan which reads only the first
+    /// two columns by using the schema `[a, b]`.
+    ///
+    /// [`Schema`]: crate::schema::Schema
+    pub fn with_schema(mut self, schema: SchemaRef) -> Self {
+        self.schema = Some(schema);
+        self
+    }
+
+    /// Optionally provide a [`SchemaRef`] for columns to select from the [`TableChanges`]. See
+    /// [`TableChangesScanBuilder::with_schema`] for details. If `schema_opt` is `None` this is a no-op.
+    pub fn with_schema_opt(self, schema_opt: Option<SchemaRef>) -> Self {
+        match schema_opt {
+            Some(schema) => self.with_schema(schema),
+            None => self,
+        }
+    }
+
+    /// Optionally provide an expression to filter rows. For example, using the predicate `x <
+    /// 4` to return a subset of the rows in the scan which satisfy the filter. If `predicate_opt`
+    /// is `None`, this is a no-op.
+    ///
+    /// NOTE: The filtering is best-effort and can produce false positives (rows that should should
+    /// have been filtered out but were kept).
+    pub fn with_predicate(mut self, predicate: impl Into<Option<ExpressionRef>>) -> Self {
+        self.predicate = predicate.into();
+        self
+    }
+
+    /// Build the [`TableChangesScan`].
+    ///
+    /// This does not scan the table at this point, but does do some work to ensure that the
+    /// provided schema make sense, and to prepare some metadata that the scan will need.  The
+    /// [`TableChangesScan`] type itself can be used to fetch the files and associated metadata required to
+    /// perform actual data reads.
+    pub fn build(self) -> DeltaResult<TableChangesScan> {
+        // if no schema is provided, use snapshot's entire schema (e.g. SELECT *)
+        let logical_schema = self
+            .schema
+            .unwrap_or(self.table_changes.schema.clone().into());
+        let mut have_partition_cols = false;
+        let mut read_fields = Vec::with_capacity(logical_schema.fields.len());
+
+        // Loop over all selected fields and note if they are columns that will be read from the
+        // parquet file ([`ColumnType::Selected`]) or if they are partition columns and will need to
+        // be filled in by evaluating an expression ([`ColumnType::Partition`])
+        println!("Logical schema: {:?}", logical_schema);
```

Comment:
```suggestion
```

Reply:
(or `debug!` if we think it's helpful?)

Reply:
Ah sry this was from testing

Reply:
I don't think we need to print on every logical schema. Deleted the line and marking this resolved.